packages/turbo-codemod/src/commands/migrate/index.ts

```diff
@@ -0,0 +1,197 @@
+  });
+  if (!exists) {
+    return endMigration({
+      success: false,
+      message: `Directory ${chalk.dim(`(${root})`)} does not exist`,
+    });
+  }
+
+  if (!looksLikeRepo({ directory: root })) {
+    return endMigration({
+      success: false,
+      message: `Directory (${chalk.dim(
+        root
+      )}) does not appear to be a repository`,
+    });
+  }
+
+  // step 1
+  const fromVersion = getCurrentVersion(selectedDirectory, options);
+  if (!fromVersion) {
+    return endMigration({
+      success: false,
+      message: `Unable to infer the version of turbo being used by ${directory}`,
+    });
+  }
+
+  // step 2
+  const toVersion = await getLatestVersion(options);
+  if (!toVersion) {
+    return endMigration({
+      success: false,
+      message: `Unable to fetch the latest version of turbo`,
+    });
+  }
+
+  if (fromVersion === toVersion) {
+    return endMigration({
+      success: true,
+      message: `Nothing to do, current version (${chalk.bold(
+        fromVersion
+      )}) is the same as the requested version (${chalk.bold(toVersion)})`,
+    });
+  }
+
+  // step 3
+  const codemods = await getCodemodsForMigration({ fromVersion, toVersion });
+  if (codemods.length === 0) {
+    console.log(
+      `No codemods required to migrate from ${fromVersion} to ${toVersion}`,
+      os.EOL
+    );
+  }
+
+  // step 4
+  console.log(
+    `Upgrading turbo from ${chalk.bold(fromVersion)} to ${chalk.bold(
+      toVersion
+    )}`,
+    os.EOL
+  );
+  const results = codemods.map((codemod, idx) => {
+    console.log(
+      `(${idx + 1}/${codemods.length}) ${chalk.bold(
+        `Running ${codemod.value}`
+      )}`
+    );
+
+    const result = codemod.transformer({ root: selectedDirectory, options });
+    Runner.logResults(result);
+    return result;
+  });
+
+  const hasTransformError = results.some(
+    (result) =>
+      result.fatalError ||
+      Object.keys(result.changes).some((key) => result.changes[key].error)
+  );
+
+  if (hasTransformError) {
+    return endMigration({
+      success: false,
+      message: `Could not complete migration due to codemod errors. Please fix the errors and try again.`,
+    });
+  }
+
+  // step 5
+  const upgradeCommand = getTurboUpgradeCommand({
+    directory: selectedDirectory,
+    to: options.to,
+  });
+
+  if (!upgradeCommand) {
+    return endMigration({
+      success: false,
+      message: "Unable to determine upgrade command",
+    });
+  }
+
+  if (options.install) {
+    if (options.dry) {
+      console.log(
+        `Upgrading turbo with ${chalk.bold(upgradeCommand)} ${chalk.dim(
+          "(dry run)"
+        )}`,
+        os.EOL
+      );
+    } else {
+      console.log(`Upgrading turbo with ${chalk.bold(upgradeCommand)}`, os.EOL);
+      execSync(upgradeCommand, { cwd: selectedDirectory });
+    }
+  } else {
+    console.log(`Upgrade turbo with ${chalk.bold(upgradeCommand)}`, os.EOL);
+  }
+
+  endMigration({ success: true });
+}
```

Comment:
It would be somewhat cool to generate a patch at the end of each migration. Might get a little hairy to get that intermediate diffs if they're all going to be unstaged. Could move to a new branch and do `git commit` in between each step also. The main idea being that you want to see the change each codemod made, in addition to the sum total. Might also help _us_ field debug requests.

Reply:
By default, this will already show you a condensed table view of the files modified, with the additions and deletions _per transform_. You can also pass in `--print` to optionally see the diffs of the files themselves for each transform.

Reply:
A temp branch is a cool idea, I haven't seen any other codemod tools do that.

Reply:
> see the diffs of the files themselves for each transform.

oh that's cool, if that's possible. Is it a full diff _per_ transform? Are you storing somewhere in between each transform? (no need to point to code, just confirming!)

Reply:
Here's the output without print:
https://app.warp.dev/block/EDiMe0dRf4ChltYW9tm71a

And here's the same run, with print:
https://app.warp.dev/block/kfIHdgynjlfIPNR63MJiPj

Reply:
awesome 👍🏾